[로또 1단계] 메다 미션 제출합니다.

medAndro opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

- domain.value에 있는 원시값을 포장한 클래스들은 init블럭에 검증 로직이 존재해서 필요성을 느꼈는데 Lotto 객체들의 묶음 검증로직은 불필요하고 생각하여 현재는 List<Lotto>를 사용했습니다. 추후 확장성을 위해 처음부터 Lottos와 같은 클래스를 두는게 좋을지 궁금합니다.
- 일부 코드에서 함수 체이닝이 과도하게 들어가지 않았나 생각되는데 메서드로 분리하는게 좋을지 아니면 현재와 같은 체이닝을 쓰는게 더 나은것인지 각각의 장단점이 있을것 같아 고민되었습니다.

jiyuneel committed:
chore: .gitignore에 local.properties 추가

jiyuneel committed:
docs: 기능 목록 작성

jiyuneel committed:
chore: .gitkeep 파일 삭제

jiyuneel committed:
feat: 1에서 45 사이의 숫자를 갖는 로또 번호 객체 구현

jiyuneel committed:
feat: 6개의 로또 번호를 갖는 로또 객체 구현

jiyuneel committed:
feat: 로또 번호 중복 여부 검사

jiyuneel committed:
feat: 6개의 당첨 번호와 보너스 번호를 갖는 당첨 로또 객체 구현

jiyuneel committed:
feat: 당첨 로또와 보너스 번호의 중복 여부 검사

jiyuneel committed:
feat: 로또 번호와 당첨 로또의 일치하는 번호의 개수를 구하는 기능 구현

jiyuneel committed:
feat: 로또 번호와 보너스 번호의 일치 여부를 판단하는 기능 구현

jiyuneel committed:
feat: 로또 등수를 구하는 로직 구현

jiyuneel committed:
feat: 로또 등수별 로또 개수를 구하는 기능 구현

jiyuneel committed:
feat: 로또 머신에서 6개의 로또 번호를 생성

jiyuneel committed:
feat: 구입 금액이 1000원 단위인지 확인하는 기능 구현

jiyuneel committed:
feat: 구입 금액이 1000원 이상인지 확인하는 기능 구현

jiyuneel committed:
feat: 구입 금액에 해당하는 만큼의 로또를 발급하는 기능 구현

jiyuneel committed:
docs: 기능 목록 수정

jiyuneel committed:
chore: package 수정

jiyuneel committed:
feat: 0 이상의 숫자인 구매 금액 구현

jiyuneel committed:
refactor: 구매 수량 객체 구현

jiyuneel committed:
feat: 수익률 객체 구현

jiyuneel committed:
feat: 로또 번호가 오름차순으로 정렬되어 있는지 검증

jiyuneel committed:
feat: 로또 생성기 객체 구현

jiyuneel committed:
feat: 구매 수량 만큼 로또를 발급하는 로또 발매기 구현

jiyuneel committed:
feat: 로또 등수별 당첨 횟수를 구하는 기능 구현

jiyuneel committed:
fix: 수익률 검증 로직 수정

jiyuneel committed:
feat: 수익률 소수점 아래 2자리로 반올림

jiyuneel committed:
feat: 당첨 정보를 바탕으로 수익률 계산

jiyuneel committed:
feat: 로또 당첨 계산기에서 당첨 정보 계산

jiyuneel committed:
chore: 로또 묶음 객체 삭제

jiyuneel committed:
refactor: 로또 번호를 구하는 컬렉션 함수 변경

jiyuneel committed:
refactor: 패키지 구조 변경

medAndro committed:
chore: 로또 패키지로 수정

medAndro committed:
feat: 구입금액 입력 로직 작성

medAndro committed:
feat: 구매 수량 출력 로직 작성

medAndro committed:
func: 구매한 로또 번호들 출력 코드 작성

medAndro committed:
feat: 당첨 로또 정보 입력 로직 구현

medAndro committed:
feat: 당첨 통계 출력 로직 작성

medAndro committed:
feat: 수익률 출력 로직 작성

medAndro committed:
feat: 보너스볼 일치여부 출력 로직 작성

medAndro committed:
chore: 당첨 통계 줄바꿈 추가

medAndro committed:
refactor: 당첨 번호 입력 로직 수정

- 입력받는곳에서 무작위 입력과 , 사이 공백 허용

medAndro committed:
refactor: 컨트롤러 내부 로직 분리

medAndro committed:
feat: 수익률 출력시 이익 손해 정보 텍스트 추가

medAndro committed:
feat: 예외 발생시 안내 메시지 추가

medAndro committed:
chore: 테스트 패키지 구조 수정

medAndro committed:
chore: 중복된 번호 예외 메시지 구체화

lee-ji-hoon commented on src/main/kotlin/lotto/domain/model/Lotto.kt:
```diff
@@ -0,0 +1,37 @@
+import lotto.constants.LottoConstants
+import lotto.domain.value.LottoNumber
+
+class Lotto(
+    val lottoNumbers: List<LottoNumber>,
+) {
```
특정 자료구조를 사용하면 외부에 중복이 불가능한 것을 알려주기도 하면서 중복에 대한 것을 자연스럽게 방지할 수가 있는데요. 그렇게 되면 init에 체이닝도 조금 줄어들 수 있겠네요

medAndro replied:
아...! Set을 사용하면 확실히 개선될것 같네요...!

lee-ji-hoon commented on src/main/kotlin/lotto/domain/model/Lotto.kt:
```diff
@@ -0,0 +1,37 @@
+                lottoNumbers.size,
+            )
+        }
+        require(lottoNumbers.map { it.number }.distinct().size == LottoConstants.NUMBER_OF_LOTTO_NUMBERS) {
+            HAS_DUPLICATE_NUMBERS.format(
+                lottoNumbers
+                    .map { it.number }
+                    .groupBy { it }
+                    .filter { it.value.size > 1 }
+                    .keys
+                    .toList()
+                    .joinToString(", "),
+            )
+        }
```
> 일부 코드에서 함수 체이닝이 과도하게 들어가지 않았나 생각되는데 메서드로 분리하는게 좋을지 아니면 현재와 같은 체이닝을 쓰는게 더 나은것인지 각각의 장단점이 있을것 같아 고민되었습니다.

위 말씀이 이 내용 같아 보이는데요~!

저는 함수 체이닝의 관심사에 따라서 함수로 분리하기도 합니다.

---

지금의 체이닝은 이해하기가 좀 어려운 감이 있습니다.

지금 한 곳에 서로 다른 역할을 하는 체이닝이 있기 때문인데요. 이 체이닝이 각각 무슨 역할을 하는지를 직접 하나하나 분석을 해서 보는 것 보다 함수로 분리를 한다면 어떨까요?

```kotlin
fun validate(lottoNumbers: List<LottoNumber>) {
    val numbers = lottoNumbers.map { it.number }
    val duplicateNumbers = findDuplicateNumbers(numbers)
    
    require(duplicateNumbers.isEmpty()) {
        HAS_DUPLICATE_NUMBERS.format(duplicateNumbers.joinToString(", "))
    }
}

private fun findDuplicateNumbers(numbers: List<Int>): List<Int> {
   ...
}

```

---

그리고 `toList()` 또한 필요 없겠네요~

medAndro replied:
제가 보기에도 확실히 이해하기 여려운 감이 느껴져서 남겨보았는데 함수로 분리하여 가독성을 높여보도록 해보겠습니다. "객체 지향 생활 체조 원칙 - 한 줄에 점을 하나만 찍는다" 의 의도가 이것인것 같네요!

lee-ji-hoon commented on src/main/kotlin/lotto/constants/LottoConstants.kt:
```diff
@@ -0,0 +1,9 @@
+package lotto.constants
+
+object LottoConstants {
+    const val LOTTO_PRICE = 1000
+    const val NUMBER_OF_LOTTO_NUMBERS = 6
+    private const val MINIMUM_LOTTO_NUMBER = 1
+    private const val MAXIMUM_LOTTO_NUMBER = 45
+    val LOTTO_RANGE = MINIMUM_LOTTO_NUMBER..MAXIMUM_LOTTO_NUMBER
+}
```
지금은 로또와 같은 작은 도메인이라 상관이 없었을수도 있지만 이것들이 계속 모이다보면 문제가 생길수도 있는데요. 

가령 `로또의 가격은 1000원` 이라는 내용은 비즈니스적인 내용으로 생각할수도 있지 않을까요?

medAndro replied:
확실히 `const val LOTTO_PRICE = 1000` 같은 경우는 가격 인상 등 서비스 운영 과정에서 유동적으로 변경되는 `비즈니스 정책`이 될 것이고, 나머지는 운영 중 변경을 할 일이 거의 없는 `도메인 규칙`을 실수로 혼재해 버린 것 같습니다. 다음 리팩터링에서는 이것을 해결해보는 노력을 해보겠습니다!

lee-ji-hoon commented on src/main/kotlin/lotto/domain/value/LottoNumber.kt:
```diff
@@ -0,0 +1,15 @@
+
+import lotto.constants.LottoConstants
+
+class LottoNumber(
+    val number: Int,
+) {
```
`LottoNumber` 를 사용하는 `Lotto` 가 수십만개 생겼다고 했을 때 `LottoNumber`도 수십만개가 생길텐데요. 이런 상황에서 원시값을 래핑한 경우 특정 클래스를 사용한다면 조금 더 최적화가 가능한데요 한 번 찾아보면 어떨까요?

medAndro replied:
리뷰를 읽고 다른 크루들이 value class를 쓰던것이 생각나서 잠깐 찾아봤더니 value class는 컴파일시 값을 직접 사용하는 방법으로 객체 생성의 오버헤드를 줄여준다는 걸 알게 되었습니다! value class를 좀 더 공부 해봐야겠네요!

lee-ji-hoon commented on src/main/kotlin/lotto/domain/value/PurchaseAmount.kt:
```diff
@@ -0,0 +1,26 @@
+
+import lotto.constants.LottoConstants.LOTTO_PRICE
+
+class PurchaseAmount(
```
`구매가격` 만으로 보이지만 init을 보면 이 클래스는 로또라는 도메인에 많이 의존되어 있는 것 같은데요.

1. 로또에 대한 의존을 줄인다.
2. 로또에 대한 의존을 두고 클래스명을 수정한다.

이렇게 2가지 방법이 존재할거 같은데 어떤 것이 적합할지 한 번 고민해보시면 좋을거 같습니다.

medAndro replied:
2번이 맞는것 같습니다. 단순히 `구매가격`이라고 한다면 0원 이상인지만 검사하는 로직만 넣었어야먄 할 것 같은 이름 같아 보이겠네요... 

`로또구매가격`인 값을 포장하는 클래스는 필요하다고 생각되어 클래스 이름만 변경하여 유지하는게 좋을것 같습니다.

lee-ji-hoon commented on src/main/kotlin/lotto/domain/value/PurchaseQuantity.kt:
```diff
@@ -0,0 +1,21 @@
+class PurchaseQuantity(
+    purchaseAmount: PurchaseAmount,
+) {
+    val quantity =
+        runCatching { purchaseAmount.amount / LottoConstants.LOTTO_PRICE }.getOrElse {
+            throw IllegalArgumentException(
+                INDIVISIBLE_VALUE_ERROR.format(
+                    purchaseAmount.amount,
+                ),
+            )
+        }
```
이 도메인 모델에 대한 테스트 커밋이 안보이네요 🤔 

이 runCatching이 정말 필요했던 것인지 한 번 테스트 코드를 구현해보면 좋을거 같습니다.

medAndro replied:
헉...다시 보니  `PurchaseAmount`에서 이미 확인했던걸 또 확인하고 있었네요

lee-ji-hoon replied:
네 맞습니다 저희는 일급객체인 `PurchaseAmount`로 이미 확인을 했으며 `PurchaseQuantity`는 그 객체를 사용하니 굳이 확인 할 필요가 없죠 👍

lee-ji-hoon commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,56 @@
+import lotto.domain.value.PurchaseAmount
+import kotlin.runCatching
+
+class InputView {
+    fun readPurchaseAmount(): PurchaseAmount {
```
View가 직접 Model을 만들어줘도 상관이 없을까요?

View의 역할은 무엇일까요? 🤔

medAndro replied:
> View 내부에는 Model의 코드가 있을 수 있고 Controller의 코드가 있으면 안된다. 

라고 단순하게 이해하고 있었고 원시 값을 포장하는 정도는 입력 뷰에서 맡아서 해도 상관 없다고 생각했었습니다. 
리뷰를 보고 자료를 찾아보면서 생각해보니 컨트롤러와 뷰에서 둘 다 도메인을 생성하여 역할 구분이 모호한 것 같습니다. 

뷰: 단순 형식 변환 (String → Int) + 도메인 객체를 전달 받아 출력 + ~~도메인 객체 생성~~
컨트롤러: 도메인 객체 생성 + 흐름 관리 + 예외 "처리"
도메인: 비즈니스 로직 구현 + 자기 검증 및 예외 "발생"

이 각각의 역할인 것 같습니다.

lee-ji-hoon commented on src/main/kotlin/lotto/domain/model/LottoWinningStats.kt:
```diff
@@ -0,0 +1,24 @@
+        return EarningRate(rate)
+    }
+
+    private fun calculateWinningAmount(): Int {
+        var result = 0
+        winningStats.forEach { (rank, count) ->
+            result += rank.winningMoney * count
+        }
+        return result
+    }
```
총 합을 구하는 것 같은데 `sumOf`를 잘 활용해보면 더 간략하게 그리고 명확하게 변경이 가능해보이네요~!

medAndro replied:
저 부분 작성하면서 분명 있을 것 같지만 생각이 안나서 미뤘었는데 `sumOf`였군요😅

lee-ji-hoon commented on src/main/kotlin/lotto/domain/service/LottoCalculator.kt:
```diff
@@ -0,0 +1,34 @@
+        return LottoWinningStats(winningStats, purchaseAmount)
+    }
+
+    private fun getWinningStats(): Map<Rank, Int> {
+        val winningStats: MutableMap<Rank, Int> =
+            Rank.entries.associateWith { 0 }.toMutableMap()
+        lottos.forEach {
+            val rank = winningLotto.getRank(it)
+            winningStats[rank] = (winningStats[rank]?.plus(1)) ?: 1
+        }
+        return winningStats
+    }
```
Mutable한 Map이 없이 이 함수가 하는 동작을 충분히 수정이 가능한데요.

체이닝을 이전에도 잘 해주셨으니 몇 가지 힌트만 드려보겠습니다.

`groupingBy` + `eachCount` 등을 잘 사용해보시면 좋을거 같습니다.

medAndro replied:
groupingBy가 리스트의 개수만 저장해서 Grouping 객체로 저장해두고 eachCount가 그룹의 개수 Value로 하는 Map으로 만드는거군요
```kt
 lottos.map { winningLotto.getRank(it) }.groupingBy { it }.eachCount()
```
이렇게 불변성도 보장하면서 한줄로 처리가 가능하겠네요...!

lee-ji-hoon replied:
함수형과 관련해서 리뷰를 남길까 고민을 했었는데 체이닝을 고민하시고 계시길래 남겨봤던 것인데 잘 반영해주셨습니다.

함수형의 경우 나중 미션에서도 계속 할테니 지금처럼 인지만 하고 계시면 좋을거 같습니다~!

lee-ji-hoon commented on src/test/kotlin/lotto/service/LottoGeneratorTest.kt:
```diff
@@ -0,0 +1,15 @@
+import org.junit.jupiter.api.assertDoesNotThrow
+
+class LottoGeneratorTest {
+    @Test
+    fun `6개의 로또 번호를 생성한다`() {
+        assertDoesNotThrow {
+            val lottoGenerator = LottoGenerator()
+            lottoGenerator.generateLotto()
+        }
+    }
```
이 테스트 코드는 함수명에 맞는 테스트 코드를 만들고 수행중인게 맞을까요?

medAndro replied:
```md
### 자동 로또 생성기
- [x] 랜덤한 `로또 번호` 객체로 이루어진 `로또` 인스턴스를 반환한다.
```
LottoGenerator 객체가 .generateLotto()를 통해 Lotto 인스턴스를 잘 만드는지 검사하고 싶었던 것이었는데 확실히 함수명이 모호하고 테스트도 단순히 예외가 발생하는지만 검사하고 있네요

함수명대로 6개의 로또 번호를 생성한다를 검사하고 싶다면
assertThat(lotto.numbers).hasSize(6)가 되어야 할텐데 이건 lotto의 테스트 코드에서 이미 검사하고 있을 것이구요

로또 인스턴스를 반환한다는 내용은
```kt
assertThat(LottoGenerator().generateLotto()).isExactlyInstanceOf(Lotto::class.java)
```
로 검사가 가능할텐데 랜덤한 `로또 번호` 객체로 이루어진 객체인지는 검사가 쉽지 않아보이네요... 🤔

lee-ji-hoon replied:
사실 `로또 인스턴스를 반환` 한다라는 것은 `RandomLottoGenerator` 에서 메인으로 봐야할 로직이 아닌거 같아요 🤔

단순 `로또 인스턴스` 라면 이것도 말씀하신대로 `Lotto()` 랑 동일한건데 여기서 확인해야 할 것은 `랜덤한 로또 번호를 포함한 Lotto를 생성 했는지`일거 같아요!

lee-ji-hoon commented on src/main/kotlin/lotto/domain/service/LottoGenerator.kt:
```diff
@@ -0,0 +1,17 @@
+import lotto.domain.model.Lotto
+import lotto.domain.value.LottoNumber
+
+class LottoGenerator {
+    fun generateLotto(): Lotto {
+        val lottoNumbers =
+            (LottoConstants.LOTTO_RANGE)
+                .shuffled()
+                .take(LottoConstants.NUMBER_OF_LOTTO_NUMBERS)
+                .sorted()
+                .map { LottoNumber(it) }
+        return Lotto(lottoNumbers)
+    }
+}
```
Lotto의 생성자나 invoke에서 처리할 수 있을거 같은데 `LottoGenerator`로 분리를 했을 때 Lotto가 해야 할 일을 나눈 것이 되는데 이렇게 됐을 때 어떤 이점이 있다고 생각하시나요?

medAndro replied:
로또를 랜덤 번호로 자동 구매하는것이 아닌 수동으로 로또번호를 선택하여 구매하는 등 Lotto를 Generate하는 기능의 추가 요구사항과 같은 확장을 염두해두고 LottoGenerator의 구현만 수정하면 Lotto 구조를 변경하지 않아도 되는 이점을 염두해 두고 분리하여 구현해보았습니다!

lee-ji-hoon replied:
그러면 이런 Lotto를 만드는 `공장`을 꼭 따로 클래스로 분리를 하는게 좋을까요?

가령 누군가 `Lotto` Class만 봤을 때 `LottoGenerator` 의 유무를 알 수 있을까요? 그냥 `Lotto` 클래스를 그냥 넣어서 만들어버릴수도 있을거 같아요.

그런 상황에서는 `Lotto` 내부에 본인이 생성을 전담하는 방법도 있는데요! 이 방법은 그냥 고민만 해봐주셔도 됩니다! 만약 제 의도가 이해가 안 된다면 말씀부탁드릴게요~!

medAndro replied:
인터페이스를 통해 별도 클래스를 두는것이 아닌 Lotto객체의 주 생성자에 Private을 걸고 Companion object에 본인 스스로를 생성시키는 메서드를 작성해서 Factory 패턴을 적용하는 방법이 있다는걸 알게되어 적용해봤습니다!

lee-ji-hoon commented on src/test/kotlin/lotto/value/EarningRateTest.kt:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(doubles = [0.051, 0.052, 0.053, 0.054])
+    fun `소수점 아래 2자리로 반올림한다`(rate: Double) {
+        val earningRate = EarningRate(rate)
+        assertThat(earningRate.rate).isEqualTo(0.05)
+    }
```
소수점 아래 2자리로 반올림한다. 이것은 어떠한 것을 확인하기 위한 테스트인가요?

medAndro replied:
`수익률은 소수점 3번째 자리에서 반올림된다` 정도를 의도하였으나 테스트명이 주어가 없어 모호하기도 하고 경계값도 제대로 테스트 하지 않았던 것 같습니다... 테스트 코드부터 다시 작성하고 재구현 해보겠습니다.

lee-ji-hoon replied:
네 좋습니다!

비즈니스적인 정보가 담긴 Model을 테스트 하시는 것이라면 말씀하신 `수익률은` 정도만 들어가도 의미가 있어 보이네요 👍

lee-ji-hoon commented on src/test/kotlin/lotto/value/PurchaseQuantityTest.kt:
```diff
@@ -0,0 +1,46 @@
+
+    @ParameterizedTest
+    @ValueSource(ints = [1000, 2000, 3000, 4000, 5000])
+    fun `구입 금액은 1000원 이상이다`(amount: Int) {
```
테스트 코드 함수명을 짤 때 `given` `when` `then` 형식으로 구현을 해보면서 이 함수명의 then은 무엇을 의미하는 것인지 생각해보면 좋을거 같네요!

medAndro replied:
이전 리뷰에서 이미 언급했듯 PurchaseAmount가 1000원 단위임을 이미 검사하고 있어서 PurchaseQuantity 테스트는 정상적인 나눠지는지만 검사하는데 관심을 가져야 했을것 같네요
```kt
// Given
val purchaseAmount = PurchaseAmount(3000)
// When
val quantity = PurchaseQuantity(purchaseAmount)
// Then
assertThat(quantity.value).isEqualTo(3)
```
와 같이 given when then 형식을 사용해서 실제 기능에 대해 테스트하는 쪽으로 고민해 보겠습니다!

lee-ji-hoon commented on src/main/kotlin/lotto/domain/model/Lotto.kt:
```diff
@@ -0,0 +1,37 @@
+import lotto.domain.value.LottoNumber
+
+class Lotto(
+    val lottoNumbers: List<LottoNumber>,
```
> domain.value에 있는 원시값을 포장한 클래스들은 init블럭에 검증 로직이 존재해서 필요성을 느꼈는데 Lotto 객체들의 묶음 검증로직은 불필요하고 생각하여 현재는 List를 사용했습니다. 추후 확장성을 위해 처음부터 Lottos와 같은 클래스를 두는게 좋을지 궁금합니다.

```
domain.value에 있는 원시값을 포장한 클래스들은 init블럭에 검증 로직이 존재해서 필요성을 느꼈는데
```

라고 해주셨는데요. 그럼 아래와 같이 `최소 하나 이상의 Lotto 티켓이 필요하다.` 라는 검증 로직이 추가가 된다고 하면 어떻게 생각하시나요?
우선 각 도메인 로직이 캡슐화가 된다는 것에 의미가 생길텐데요.

```kotlin
data class Lottos(val tickets: List<Lotto>) {
    init {
        require(tickets.isNotEmpty()) { "최소 하나 이상의 Lotto 티켓이 필요합니다." }
    }
    
    // 예시로 전체 티켓을 정렬하는 기능
    fun sortedTickets(): List<Lotto> = tickets.sortedBy { it.lottoNumbers.first() }
    
    // 추후 추가될 수 있는 Lotto 집합 관련 비즈니스 로직 구현 가능
}
```

도메인 모델의 의의는 `검증 로직`이 아닌 각 객체가 갖는 `도메인적인 비즈니스` 에 의의를 두고 고민을 해보시는게 좋을거 같습니다.

medAndro replied:
도메인적인 비즈니스를 염두하여 고민해 보았을 때 `로또 티켓 집합`에 대한 당첨 통계의 계산, 수익률의 계산을 `Lottos`에서 맡게 하여 현재 컨트롤러에서 List<Lotto>를 만들어 각 클래스에게 직접 시키고 있던 것을 `Lottos`에게 시키는 방향으로 바꾸는것이 좋을 것 같다고 생각해서 이러한 방향으로 수정해 보겠습니다

lee-ji-hoon reviewed (changes requested):
안녕하세요 메다! 로또 리뷰어 코니입니다!

전반적으로 잘 구현을 해주셨는데요. View의 관점과 그리고 Domain Model 에서 챙겨볼만한 것들이 몇가지 있어서 리뷰를 남겨봤는데요! 혹시나 이해가 너무 안 되는 부분이 있다면 편하게 연락 해주세요!

로또 미션 화이팅입니다!

medAndro committed:
refactor: getWinningStats() 메서드 로직 개선

- MutableMap을 제거하여 불변성 보장 및 가독성 향상

medAndro committed:
refactor : LottoGenerator 테스트 코드 수정

medAndro committed:
refactor: LottoNumber의 내부 자료형을 Set으로 변경

- 불필요해진 중복검사 테스트 코드와 검사 로직 제거
- 로또 번호들의 Set을 정렬된 리스트로 반환하는 메서드 추가

medAndro committed:
refactor: LottoNumber를 value class로 변경하여 최적화 향상

medAndro committed:
refactor: 로또 구매 정보를 LottoPayInfo로 통합

- PurchaseAmount, Quantity를 통합하여 관리
- PurchaseAmount와 같이 모호했던 클래스명을 구체적 프로퍼티,함수명으로 수정

medAndro committed:
refactor: 수익률 테스트명 구체화 및 경계값으로 검사

medAndro committed:
refactor: WinningAmount 계산 코드 내부 코드 개선

- sumOf 활용으로 가변 변수 제거 및 속도 향상

medAndro committed:
refactor: LottoConstnts 내부에 혼재된 도메인 규칙을 해당 클래스로 이동

medAndro committed:
refactor: List<Lotto>를 가지는 Lottos 객체 도입

- LottoCalculator의 비즈니스 로직을 Lottos에게 이관

medAndro committed:
refactor: View의 Model 생성 역할 제거

- 최소한의 원시값 변환만 수행 후 컨트롤러에서 모델 생성을 담당하도록 변경

medAndro committed:
refactor: LottoGenerator 구현 수정

- RandomLottoGenerator로 의미 명확화
- 메소드 체이닝을 함수로 분리

medAndro committed:
feat: 0번 당첨을 포함한 당첨 통계 정보를 반환하는 로직 추가

medAndro committed:
refactor: assertThat을 사용한  테스트들을 given when then 형식으로 수정

medAndro committed:
docs: 수정된 기능 요구사항 README 반영

medAndro committed:
chore: Lotto의 예외 메시지 수정

medAndro commented:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 코니! 꼼꼼하게 작성해주신 리뷰들 덕분에 공부가 많이 된 것 같습니다...!

코드를 건드리기에 앞서 리뷰주신부분을 순서대로 공부해가며 우선적으로 제 생각을 정리해 리뷰마다 ❤️이모지와 함께 답글을 달았고 답글로 작성한대로 리팩터링을 하며 🚀 이모지를 달았으며 아래와 같은 내용을 반영했습니다.

- Lotto의 lottoNumbers 자료구조를 Set으로 변경
- 과도한 메서드 체이닝을 함수로 분리
- LottoConstants 내부의 변경 가능성이 적은 도메인 규칙을 각각의 객체로 이관
- LottoNumber를 일반 클래스에서 Value 클래스로 변경하여 최적화
- PurchaseAmount와 PurchaseQuantity를 LottoPayInfo로 통합
- InputView에서 객체를 직접 반환하는것이 아닌 최소한의 변환만 수행한 값을 반환 
- sumOf를 활용하여 당첨 금액을 계산하여  forEach를 통한 가변 변수 접근 제거
- groupingBy와 eachCount를 사용하여 Mutable한 Map이 없이 당첨 통계 작성
- LottoGenerator와 EarningRateTest의 테스트명 구체화 및 재작성
- assertThat를 사용한 테스트코드를 given when then 형식으로 변경
- List<Lotto>를 프로퍼티로 가지는 Lottos 객체 도입

제가 의도를 잘 파악하여 잘 반영하였는지, 제 스스로 코틀린 문법과 MVC 패턴, 객체지향적 사고에 아직 미숙하다고 생각하고있어서 혹시라도 제가 잘못 이해하거나 사용중인 부분이 있는지 확인해주시면 감사하겠습니다!

lee-ji-hoon commented on src/main/kotlin/lotto/view/OutputView.kt:
```diff
@@ -0,0 +1,34 @@
+        }
+    }
+
+    fun printLottoEarningRate(earningRate: EarningRate) {
+        val gainLossText = if (earningRate.rate > 1) "이득이라는" else (if (earningRate.rate == 1.0) "본전이라는" else "손해라는")
+        println("총 수익률은 ${earningRate.rate}입니다.(기준이 1이기 때문에 결과적으로 $gainLossText 의미임)")
+    }
```
지금까지는 TDD로 만들어진 Model 관점에서 리뷰를 많이 남겼는데요. 

이번엔 View에 대해서 고민을 한 번 해보면서 다음 미션을 진행해보면 좋을거 같네요!

View의 관점에서는 지금 1보다 큰지 작은지 같은지를 비교 하기 보다는 rate에 따른 `상태` 를 만들고 그 상태에 따라 Text가 변경이 되는건 어떨까요?

medAndro replied:
확실히 수익률에서 손익 정보는 View에서 판단하는것이 아닌 손익 상태를 직접 넘겨주는게 View의 관점에서 맞는것 같아 보이네요!

추가로 당첨 통계 출력 부분에서 보너스볼 일치 상태도 추가해야 할까 생각하다가
```kt
val bonusText = if (winningStats.key == Rank.SECOND) ", 보너스 볼 일치" else " "
```
Rank.SECOND가 이미 보너스볼이 일치한다는걸 암시한다고 생각해서 상태를 추가로 만들 필요가 없을거 같아 당첨 통계 출력은 그대로 놔두려 하는데 VIew의 관점에서 이게 맞는것인지 살짝 아리송하네요...

lee-ji-hoon reviewed (approved):
안녕하세요 메다! 로또 1단계는 여기서 마무리해도 좋을거 같네요 👍 

처음부터 구현을 잘 해주셔서 학습 범위보다 조금 더 넓은 범위도 요청을 드려봤는데 잘 반영해주셨습니다.

View에 관점에서 생각해볼만한 고민거리 하나를 남겨놨으며, 지금 남겨드린 리뷰는 2단계를 진행하면서 고민 및 반영을 해주시면 될거 같습니다 👍 

2단계도 화이팅입니다!
